go/teams/box_audit.go

```diff
@@ -0,0 +1,1104 @@
+
+var _ Versioned = &BoxAuditJail{}
+
+func (j *BoxAuditJail) GetVersion() Version {
+	return j.Version
+}
+
+func NewBoxAuditJail(version Version) *BoxAuditJail {
+	return &BoxAuditJail{
+		TeamIDs: make(map[keybase1.TeamID]bool),
+		Version: version,
+	}
+}
+
+func (a *BoxAuditor) shouldAudit(mctx libkb.MetaContext, team Team) (bool, *keybase1.BoxAuditAttemptResult, error) {
+	if team.IsSubteam() {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_SUBTEAM
+		return false, &res, nil
+	}
+	if team.IsOpen() {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_OPENTEAM
+		return false, &res, nil
+	}
+	role, err := team.MemberRole(mctx.Ctx(), mctx.CurrentUserVersion())
+	if err != nil {
+		return false, nil, err
+	}
+	if !role.IsOrAbove(keybase1.TeamRole_WRITER) {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_ROLE
+		return false, &res, nil
+	}
+
+	return true, nil, nil
+}
+
+// loadTeamForBoxAudit loads a team once, but if the client
+// has not yet stored BoxSummaryHashes (due to being an old client)
+// it does a force full reload so it is populated.
+func loadTeamForBoxAudit(mctx libkb.MetaContext, teamID keybase1.TeamID) (*Team, error) {
+	return loadTeamForBoxAuditInner(mctx, teamID, false)
+}
+
+func loadTeamForBoxAuditInner(mctx libkb.MetaContext, teamID keybase1.TeamID, force bool) (team *Team, err error) {
+	defer mctx.TraceTimed("loadTeamForBoxAuditInner", func() error { return err })()
+	arg := keybase1.LoadTeamArg{
+		ID:              teamID,
+		ForceRepoll:     true,
+		Public:          teamID.IsPublic(),
+		ForceFullReload: force,
+	}
+
+	team, err = Load(mctx.Ctx(), mctx.G(), arg)
+	if err != nil {
+		return nil, err
+	}
+	if team == nil {
+		return nil, fmt.Errorf("got nil team from loader")
+	}
+
+	// If the team sigchain state was constructed with support for box summary
+	// hashes, the map will be non-nil but empty. It will only be nil if the
+	// state is cached from a team load before box summary hash support.
+	if team.GetBoxSummaryHashes() == nil {
+		if force {
+			return nil, fmt.Errorf("failed to get a non-nil box summary map after full reload")
+		}
+		return loadTeamForBoxAuditInner(mctx, teamID, true)
+	}
+	return team, nil
+}
+
+// calculateExpectedSummary loads all the UPAKs of the team's members and
+// generates all the PUK generations we expect the team's secret boxes to currently be encrypted for.
+func calculateExpectedSummary(mctx libkb.MetaContext, team *Team) (summary *boxPublicSummary, err error) {
+	defer mctx.TraceTimed("calculateExpectedSummary", func() error { return err })()
+
+	// We use a set because a UV could appear both explicit member and as an implicit admin
+	members, err := team.Members()
+	if err != nil {
+		return nil, err
+	}
+	memberUVs := members.AllUserVersions()
+
+	uvSet := make(map[keybase1.UserVersion]bool, len(memberUVs)) // prealloc; will be at least this size
+	for _, memberUV := range memberUVs {
+		uvSet[memberUV] = true
+	}
+
+	if team.IsSubteam() {
```

Comment:
should we remove this part? it probably doesn't hurt to keep it, maybe with some additional comment / note here.

Reply:
nah, i'd rather keep this in without a comment since its correct (since I feel like we would just forget to delete the comment later)

Reply:
test

Reply:
test

Reply:
test

Reply:
test